src/metabase/models/dashboard.clj

```diff
@@ -131,6 +141,22 @@
   [dashboard]
   (update-dashboard-subscription-pulses! dashboard))
 
+(defn- card-ids-by-param-id
+  "Return a map (keyed by parameter ID) of ValuesCards associated with `dashboard`"
+  [dashboard-id]
+  (->> (db/select ValuesCard :parameterized_object_id dashboard-id :parameterized_object_type "dashboard")
+       (group-by :parameter_id)
+       (m/map-vals (comp :card_id first))))
+
+(defn- populate-values-card
+  [{dashboard-id :id parameters :parameters :as dashboard}]
+  (assoc dashboard :parameters
+         (let [param-id->card-id (card-ids-by-param-id dashboard-id)]
+           (for [{:keys [id source_type card_id] :as param} parameters]
+             (if (= source_type "card")
+               (assoc param :card_id (get param-id->card-id id card_id))
```

Comment:
why do we need to do this?

Reply:
The link between ParameterCards and Dashboards is determined by `dashboard->parameters->source_options->card_id`, which will be modified by the FE as appropriate. So the FE needs to know what the current associated `card_id` is in case there's no change.

Without having this `populate-card-id-for-parameters` in the `post-select`, `(db/insert! Dashboard (db/select Dashboard ...))` would delete the associated ParameterCard since the `select`-ed Dashboard wouldn't have a `card_id` for each parameter and then the `insert!` would delete any ParameterCards associated with the Dashboard that weren't named in `parameters`

Reply:
Isn't the card_id already in `parameters[0].source_options` ? I don't think we need to populate it at all.

Unless I'm missing something.